Change how we hoist filters from a subquery

PavelSafronov opened the pull request:
Before hoisting a filter condition inside a subquery, check if the condition is using any of the subquery tables.

This also includes a minor renaming change: when hoisting a table and there's a possibility of a name collision, we now add an unambiguous table alias.

This resolves https://github.com/dolthub/dolt/issues/5342

max-hoffman commented on enginetest/queries/queries.go:
```diff
@@ -7154,6 +7154,14 @@ With c as (
 		Expected: []sql.Row{{int64(4294967296)}},
 	},
+	{
+		Query: "select * from mytable where exists (select * from mytable where i = 1);",
```
join queries go into `join_op_tests.go` and/or `join_planning_tests.go`

PavelSafronov replied:
I think there's a benefit to having the queries in both places:
- adding a query to `queries.go` will verify that we execute the query correctly
- adding a query to `join_op_tests.go` or `join_planning_tests.go` will verify that we understand how this query is being planned

There was an alias disambiguation bug that was only caught because we executed the query: the bug wasn't manifesting if we only verified that the plan but didn't also execute the query.

max-hoffman replied:
`join_op_tests.go` will test the join with every join operator, rather than just one. Any join in `queries.go` will be more thoroughly tested in `join_op_tests.go`. The planning tests have a narrower use-case, which is that we want to make sure a specific transformation rule is applied to create a specialized join (but they also optionally test correctness).

max-hoffman reviewed (changes requested):
If i'm reading this correctly, our goal here is to unscope an EXISTS subquery in the absence of a filter bridging the inner/outer scopes. The way it is implemented here is perhaps misleading -- we are tagging the filters as movable to the outer scope even when it does not reference attributes in the outer scope. This has the side-effect of allowing a filter like `a = 1` to serve as a join condition for decorrelation. This can move filters outside of their scope of evaluation, which I think has correctness implications, for example a right-side-filter moving above a semi join.

In the general case, the pattern you are describing maybe should be converted into a CROSS join onto a MAX_1_ROW node. It is a cross join, because the filter is degenerate. And the MAX_1_ROW because we will either return all rows from the join-left (if MAX_1_ROW returns 1 row), or zero of the rows from  left (if MAX_1_ROW returns 0 rows).

PavelSafronov commented on sql/analyzer/hoist_select_exists.go:
```diff
@@ -231,11 +255,13 @@ func decorrelateOuterCols(e *plan.Subquery, scopeLen int, aliasDisambig *aliasDi
 		target, ok := nodeAliases[goodAlias]
 		if !ok {
-			return nil, fmt.Errorf("node alias %s is not in nodeAliases", goodAlias)
+			return nil, fmt.Errorf("good node alias %s is not in nodeAliases", goodAlias)
```
> We need terminology to be precise and unambiguous. What is a "good alias"?

Good point, the terminology here can be easily improved. "good" -> "unconflicted".

> How would a client know what to do with this error?

Send the stack trace/error to us? If we have this specific error text, we can instantly pinpoint where the code exploded.
I guess as long as we have a way to identify the cause of the error, that's enough. Would it be better to provide an opaque error to the client? Or panic because this is clearly an unexpected scenario?

> I really don't think we should have this check. We are performing an existence check on a list->map we created by reading from the map one line before the loop. It does not make the code easier to read.

I don't like the idea of ignoring errors, even if currently **know** that these errors are impossible. What if we introduce a bug into `findConflicts`? Assuming that the lookup will be successful will cause issues.

I had to make [this change](https://github.com/dolthub/go-mysql-server/pull/1633/files#diff-d9767d452532ba5195b8da77ef173608c57601567100f4d947bf12e513a5b7beR291) because [this lookup](https://github.com/dolthub/go-mysql-server/blob/pavel/dolt-5342/sql/analyzer/hoist_select_exists.go#L308) was failing. That was happening because we weren't properly handling `Table` ambiguity, just `TableAlias` ambiguity.

PavelSafronov commented:
This PR and hoist_select_exists rule have been rewritten almost from scratch. The below explanation is copied from the code comments.

The basic concept is that we are now considering a generic "Filter (Not) Exists" tree like this:

```
	Filter
	├─ (NOT) EXISTS Subquery
	│  └─ Filter
	│     ├─ (CONDITION)
	│     └─ RIGHT
	└─ LEFT
```

There are a few cases to consider, all around `CONDITION`. Here are the various possibilities and the equivalent plans.
Note that for each case we have plan #N (1N, 2N, 3N, 4N) for the cases where the subquery is NOT EXISTS.

	Case 1: `CONDITION` depends on `LEFT` AND `RIGHT`
		1. Need to scan both Relations
		2. Convert 1 to `SemiJoin(LEFT, RIGHT, CONDITION)`
		3. Convert 1N to `AntiJoin(LEFT, RIGHT, CONDITION)`
	Case 2: `CONDITION` depends on only `LEFT`
		1. No need to scan RIGHT, just need to know if there are any rows in RIGHT
		2. Convert 2 to `Proj(LEFT, CrossJoin(Filter(LEFT, CONDITION), Limit1(RIGHT)))`
		3. Convert 2N to `Proj(LEFT, CrossJoin(Filter(LEFT, NOT(CONDITION)), Limit1(RIGHT)))`
	Case 3: `CONDITION` depends on only `RIGHT`
		1. Need to scan both Relations
		2. Convert 3 to	`Proj(LEFT, CrossJoin(LEFT, Limit1(Filter(RIGHT, CONDITION))))`
		3. Convert 3N to `AntiJoin(LEFT, Limit1(Filter(RIGHT, CONDITION)), CONDITION)`
	Case 4: `CONDITION` DOES NOT depend on `LEFT` OR `RIGHT`
		1. No need to scan RIGHT
		2. Convert 4 to `Filter(LEFT, CONDITION)`
		3. Convert 4N to `Filter(LEFT, NOT(CONDITION))`

If hoisting results in naming conflicts, we will rename the conflicting aliases/tables in the subquery.
(Renaming tables/alias in a WHERE EXISTS subquery is perfectly safe, since the subquery results are never actually used.)

This is a large change from the previous logic, which split up the filter into multiple expressions that ended up in different spots in the tree. I think this approach is somewhat simpler, but it does have the disadvantage that it does not (yet) work on filters with multiple exists subqueries.

max-hoffman commented:
I've done a scan through, supporting both conjunction top-level filters (EXISTS AND ...) and conjunctions in the subquery scope filters is necessary to avoid customer regressions. I can try to take a closer look tomorrow if you want feedback on how to approach that or whether a different approach is a lower lift. I am close on a PR for a separate bug that adds a rule to pull filters from subqueries into parent scopes, which maybe simplifies cases (2) and (4). Do you think hitting all of those transformations is necessary to fix the immediate customer bug?

PavelSafronov commented:
> I've done a scan through, supporting both conjunction top-level filters (EXISTS AND ...) and conjunctions in the subquery scope filters is necessary to avoid customer regressions. I can try to take a closer look tomorrow if you want feedback on how to approach that or whether a different approach is a lower lift. I am close on a PR for a separate bug that adds a rule to pull filters from subqueries into parent scopes, which maybe simplifies cases (2) and (4). Do you think hitting all of those transformations is necessary to fix the immediate customer bug?

You're right, a complete rewrite was overkill. I've reworked the approach and added more tests. This change takes care of the original bug and does so without breaking things.

max-hoffman commented on sql/analyzer/hoist_select_exists.go:
```diff
@@ -166,11 +168,8 @@ func hoistExistSubqueries(scope *Scope, a *Analyzer, filter *plan.Filter, scopeL
-	}
-	log.Debug(sql.DebugString(ret))
-	return ret, transform.NewTree, nil
+	log.Debug(sql.DebugString(result))
```
debugging leftover from previous PR
```suggestion
```

PavelSafronov replied:
Removing.

max-hoffman commented on sql/analyzer/join_order_builder.go:
```diff
@@ -159,28 +159,69 @@ func (j *joinOrderBuilder) reorderJoin(n sql.Node) {
+	actionableChild := n.Child
+
+	filter, isChildFilter := actionableChild.(*plan.Filter)
+	if isChildFilter {
```
seems like we can probably just build child recursively, and then tag limit onto the group returned. if you have LIMIT->SELECT or SELECT->LIMIT the recursion works regardless. (Also SELECT -> LIMIT probably is not valid SQL?)

PavelSafronov replied:
Definitely, this is simpler than having the two separate functions:
```go
	case *plan.Filter:
		_, _, group = j.populateSubgraph(n.Child)
		group.relProps.filter = n.Expression
	case *plan.Limit:
		_, _, group = j.populateSubgraph(n.Child)
		group.relProps.limit = n.Limit
```

PavelSafronov replied:
Definitely simpler to have this than two extra functions:
```go
	case *plan.Filter:
		_, _, group = j.populateSubgraph(n.Child)
		group.relProps.filter = n.Expression
	case *plan.Limit:
		_, _, group = j.populateSubgraph(n.Child)
		group.relProps.limit = n.Limit
```

max-hoffman commented on sql/analyzer/memo.og.go:
```diff
@@ -512,50 +512,60 @@ func formatRelExpr(r relExpr) string {
+		return nil, err
+	}
+
+	result = r.group().relProps.addFilterAndLimit(result)
```
how do you feel about something like `result = r.group().finalize(result)`, though not particularly important

PavelSafronov replied:
`finalize` and moving the method to `expGroup` sounds good, making the change.

max-hoffman commented on sql/analyzer/join_order_builder.go:
```diff
@@ -159,28 +159,69 @@ func (j *joinOrderBuilder) reorderJoin(n sql.Node) {
+	return group
+}
+
+func (j *joinOrderBuilder) buildFilter(n *plan.Filter) *exprGroup {
```
I'd like to switch over to SQL terminology FILTER->SELECT for the memo if possible, so `buildFilter`, etc...

PavelSafronov replied:
These methods are being removed.

max-hoffman commented on sql/analyzer/exec_builder.go:
```diff
@@ -110,6 +110,15 @@ func (b *ExecBuilder) buildLookup(l *lookup, input sql.Schema, children ...sql.N
 			ret = plan.NewTableAlias(n.Name(), ret)
 		}
 		ret = plan.NewDistinct(ret)
+	case *plan.Filter:
```
Organizational simplification opp: if you made filter, distinct, and SelectSingleRel implement `sql.UnaryNode`, and add `UnaryNode.WithChild()`, all of these could condense to this:
```golang
	case sql.UnaryNode:
		switch n := n.Child().(type) {
		case *plan.ResolvedTable:
			ret, err = plan.NewIndexedAccessForResolvedTable(n, plan.NewLookupBuilder(l.index, keyExprs, l.nullmask))
		case *plan.TableAlias:
			ret, err = plan.NewIndexedAccessForResolvedTable(n.Child.(*plan.ResolvedTable), plan.NewLookupBuilder(l.index, keyExprs, l.nullmask))
			ret = plan.NewTableAlias(n.Name(), ret)
		}
		ret = n.WithChild(ret)
```

max-hoffman replied:
`*plan.SelectSingleRel` is also kind of defunct now, could just attach the filter to the exprGroup's select. Which would also let us delete the `normalizeSelectSingleRel` rule.

max-hoffman replied:
Neither of these are blocking for this PR

PavelSafronov replied:
I tried changing filter and distinct to implement `sql.UnaryNode`, but this requires a lot of changes to be made due to go's composition: `filter.Child` used to return `filter.UnaryNode.Child`, but now `Child` is an un-invoked function.
So every place we ever used `filter.Child`/`distinct.Child`, now have to be updated to `filter.Child()`/`distinct.Child()`. These are changes we'd have to make to both GMS and dolt.

But the worst thing is that we don't get notified of all of these cases at build time: [this particular line](https://github.com/dolthub/go-mysql-server/blob/main/sql/plan/filter.go#L92) does not generate a build-time error, since `DebugString` accepts *anything*, but this line will cause a run-time error.

So changing filter and distinct to implement `sql.UnaryNode` is quite risky. For now, I'm going to leave a todo in exec_builder with this suggestion.

I haven't tried the `*plan.SelectSingleRel`, but that one also sounds pretty involved. But I can add a todo to `normalizeSelectSingleRel` for us to consider removing it.

max-hoffman commented on sql/analyzer/aliases.go:
```diff
@@ -286,6 +288,14 @@ func renameAliases(node sql.Node, oldNameLower string, newName string) (sql.Node
 		}
 
+		// if the node is a table and its parent is not a table alias, add a table alias node
+		if _, ok := newNode.(sql.Table); ok {
```
This might be missing subquery aliases and table functions, maybe others. And nit: the plan looks a bit nicer with this, but practically speaking `TableAlias(A)->TableAlias(B)->Tablescan(C)` is equivalent to `TableAlias(A)->Tablescan(C)`. `transform.NodeWithCtx` is slightly more expensive than `transform.Node`, and the logic can be harder to follow when parent checking isn't strictly necessary. Other rules duplicate table aliases in a maybe undesirable way; having a `deduplicateTableAliases` rule that does: `(TableAlias1->TableAlias2)=>(TableAlias1)` might be useful at some point.

Something like this would maybe catch the other nameable nodes:

```golang
switch node := node {
case sql.NameableNode:
  newNode = plan.NewTableAlias(newName, node)
default:
}
```

PavelSafronov replied:
Reverted this change.

PavelSafronov replied:
PR updated. We're now directly renaming TableAlias, SubqueryAlias, and Procedure. And in case where we see a Nameable with the old name, we rename by introducing a TableAlias on top of the node.

max-hoffman commented on enginetest/memory_engine_test.go:
```diff
@@ -204,30 +204,72 @@ func TestSingleQueryPrepared(t *testing.T) {
 
 // Convenience test for debugging a single query. Unskip and set to the desired query.
 func TestSingleScript(t *testing.T) {
-	t.Skip()
```
one nit of mine that others disagree with is that i usually reset this file back to main before commiting. 1) We forget to re skip this sometimes, which is bad because this suite shouldn't be tested. And b) for most GMS PRs its kind of annoying to review a PR where this file's testing changes outweigh the relevant diff.

PavelSafronov replied:
Agreed, will revert all memory_engine_test.go changes.

max-hoffman commented on enginetest/join_planning_tests.go:
```diff
@@ -391,14 +391,14 @@ order by 1;`,
 				q: `SELECT * FROM xy WHERE (
       				EXISTS (SELECT * FROM xy Alias1 WHERE Alias1.x = (xy.x + 1))
       				AND EXISTS (SELECT * FROM uv Alias1 WHERE Alias1.u = (xy.x + 2)));`,
-				types: []plan.JoinType{plan.JoinTypeSemiLookup, plan.JoinTypeSemiLookup},
```
We should fix this before merging.

PavelSafronov replied:
Reverted this change.

PavelSafronov replied:
The reason these changes are happening, is because we're no longer hoisting the filters, and so these remain subqueries.

For this SQL:
```sql
SELECT * FROM xy WHERE (
      				EXISTS (SELECT * FROM xy Alias1 WHERE Alias1.x = (xy.x + 1))
      				AND EXISTS (SELECT * FROM uv Alias2 WHERE Alias2.u = (xy.x + 2)));
```

Here's the starting tree:
```txt
Project
 ├─ columns: [*]
 └─ Filter
     ├─ AND
     │   ├─ EXISTS Subquery
     │   │   ├─ cacheable: false
     │   │   └─ Project
     │   │       ├─ columns: [*]
     │   │       └─ Filter
     │   │           ├─ (Alias1.x = (xy.x + 1))
     │   │           └─ TableAlias(Alias1)
     │   │               └─ UnresolvedTable(xy)
     │   └─ EXISTS Subquery
     │       ├─ cacheable: false
     │       └─ Project
     │           ├─ columns: [*]
     │           └─ Filter
     │               ├─ (Alias2.u = (xy.x + 2))
     │               └─ TableAlias(Alias2)
     │                   └─ UnresolvedTable(uv)
     └─ Table
         ├─ name: xy
         └─ columns: [x y]
```

And the final tree:
```txt
QueryProcess
 └─ Filter
     ├─ AND
     │   ├─ EXISTS Subquery
     │   │   ├─ cacheable: false
     │   │   └─ Filter
     │   │       ├─ (Alias1.x = (xy.x + 1))
     │   │       └─ TableAlias(Alias1)
     │   │           └─ Table
     │   │               ├─ name: xy
     │   │               └─ columns: [x y]
     │   └─ EXISTS Subquery
     │       ├─ cacheable: false
     │       └─ Filter
     │           ├─ (Alias2.u = (xy.x + 2))
     │           └─ TableAlias(Alias2)
     │               └─ Table
     │                   ├─ name: uv
     │                   └─ columns: [u v]
     └─ Table
         ├─ name: xy
         └─ columns: [x y]
```

max-hoffman reviewed (commented):
some progress comments

max-hoffman reviewed (changes requested):
I am seeing three sets of changes here, alongside a bunch of tests:

1. limits and filters into the memo

2. changes that limit hoisting

3. disambiguate all nameables

(1) is missing a bridge into the memo builder, so it can't be doing anything right now. I confirmed locally that these changes aren't used right now.

(2) doesn't appear to be correct, and blocks valid transforms (which makes some of the tests accidentally pass b/c they aren't hitting unnesting pattern).

(3) seems like the only salient change, but i still think the version on the last pass i suggested is easier. I tested the code i pasted below and it seems to work.

I think we undo (1) and (2) either way, until the point where we actually need (1) and then flush out that implementation. Then the options are i) fix the remaining bugs, which look related to hoisting filters above their appropriate scopes when i test locally; ii) choose the subset of tests that work with the renaming change, comment out the others, and then follow-up with the NULL filter getting lifted above the tree. The remaining bugs in (i) look interesting and maybe worth their own PR. I'd recommend (ii) on a fresh repo b/c this PR is getting kind of hairy. If you get (i) to work that is cool also, but I think in general this PR is a lesson in biting off chewable chunks. The first thing you should do when fixing a query is simplify until you've repro'd one bug. After you fix that one bug, loop back to the original query, and if there are more bugs start at step 1 again.

PavelSafronov commented on sql/analyzer/aliases.go:
```diff
@@ -272,28 +274,53 @@ func renameAliasesInExp(exp sql.Expression, oldNameLower string, newName string)
-			if strings.EqualFold(tableAlias.Name(), oldNameLower) {
-				newNode = tableAlias.WithName(newName)
+		// update node directly
+		switch node := node.(type) {
+		case *plan.TableAlias:
+			if strings.EqualFold(node.Name(), oldNameLower) {
+				a.Log("renaming table alias %s to %s", node.Name(), newName)
+				newNode = node.WithName(newName)
+				allSame = transform.NewTree
+			}
+		case *plan.SubqueryAlias:
+			if strings.EqualFold(node.Name(), oldNameLower) {
+				a.Log("renaming subquery alias %s to %s", node.Name(), newName)
+				newNode = node.WithName(newName)
+				allSame = transform.NewTree
+			}
+		case *plan.Procedure:
+			if strings.EqualFold(node.Name, oldNameLower) {
+				a.Log("renaming procedure %s to %s", node.Name, newName)
+				newNode = node.WithName(newName)
+				allSame = transform.NewTree
+			}
+		}
+
+		// if parent was not updated, we may need to update children
+		if allSame == transform.SameTree {
+			nameable, ok := node.(sql.Nameable)
+			if ok && strings.EqualFold(nameable.Name(), oldNameLower) {
+				a.Log("renaming %s to %s", node, newName)
+				newNode = plan.NewTableAlias(newName, node)
 				allSame = transform.NewTree
 			}
 		}
```
Neat! Love the fact that we can "add" an interface after the fact.

PavelSafronov commented on sql/analyzer/hoist_select_exists.go:
```diff
@@ -203,9 +205,12 @@ func decorrelateOuterCols(e *plan.Subquery, scopeLen int, aliasDisambig *aliasDi
-					// has to be from out of scope
-					outerRef = true
+				if ok {
+					_, sourceInSubquery := queryAliases[gf.Table()]
```
The filters we identify as being "inner" will end up on the Filter that's outside the SemiJoin. I think if the filter is using a subquery table, we want to keep that filter on the inside of the SemiJoin, not outside, so we need to put that filter into the `outerFilters` collection.

PavelSafronov commented:
Right you are, @max-hoffman , this PR has had too many updates and it's quite hairy. I followed the steps you outlined and created a minimal PR that fixes the hoisting bug and fixes the renaming bug: https://github.com/dolthub/go-mysql-server/pull/1665

We're not doing anything with filters/limits yet, so those changes are not in the new PR.
